icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java

```diff
@@ -1951,4 +1951,89 @@ public void TestInfiniteRecursion() {
+                "NaN: not a number;\n" +
+                "zero; one; two; three; four; five; six; seven; eight; nine;\n" +
+                "ten; eleven; twelve; thirteen; fourteen; fifteen; sixteen; seventeen; eighteen; nineteen;\n" +
+                "20: twenty[->>];\n" +
+                "30: thirty[->>];\n" +
+                "40: forty[->>];\n" +
+                "50: fifty[->>];\n" +
+                "60: sixty[->>];\n" +
+                "70: seventy[->>];\n" +
+                "80: eighty[->>];\n" +
+                "90: ninety[->>];\n" +
+                "100: << hundred[ >>];\n" +
+                "1000: << thousand[ >>];\n" +
+                "1000000: << million[ >>];\n" +
+                "1000000000: << billion[ >>];\n" +
+                "1000000000000: << trillion[ >>];\n" +
+                "1000000000000000: =#,##0=;\n" +
+                "%ordinal:\n" +
+                "-x: minus >>;\n" +
+                "x.x: =#,##0.#=;\n" +
+                "Inf: infinitieth;\n" +
+                "zeroth; first; second; third; fourth; fifth; sixth; seventh; eighth; ninth;\n" +
+                "tenth; eleventh; twelfth;\n" +
+                "13: =%cardinal=th;\n" +
+                "20: twent[y->>|ieth];\n" +
+                "30: thirt[y->>|ieth];\n" +
+                "40: fort[y->>|ieth];\n" +
+                "50: fift[y->>|ieth];\n" +
+                "60: sixt[y->>|ieth];\n" +
+                "70: sevent[y->>|ieth];\n" +
+                "80: eight[y->>|ieth];\n" +
+                "90: ninet[y->>|ieth];\n" +
+                "100: <%cardinal< [$(cardinal,one{hundred}other{hundreds})$ >>|$(cardinal,one{hundredth}other{hundredths})$];\n" +
+                "1000: <%cardinal< [$(cardinal,one{thousand}other{thousands})$ >>|$(cardinal,one{thousandth}other{thousandths})$];\n" +
+                "1000000: <%cardinal< [$(cardinal,one{million}other{millions})$ >>|$(cardinal,one{millionth}other{millionths})$];\n" +
+                "1000000000: <%cardinal< [$(cardinal,one{billion}other{billions})$ >>|$(cardinal,one{billionth}other{billionths})$];\n" +
+                "1000000000000: <%cardinal< [$(cardinal,one{trillion}other{trillions})$ >>|$(cardinal,one{trillionth}other{trillionths})$];\n" +
+                "1000000000000000: =#,##0=$(ordinal,one{st}two{nd}few{rd}other{th})$;";
+        RuleBasedNumberFormat rbnf = new RuleBasedNumberFormat(rules, ULocale.US);
+
+        String[][] enTestFullData = {
+                {"20", "twentieth"},
+                {"21", "twenty-first"},
+                {"29", "twenty-ninth"},
+                {"30", "thirtieth"},
+                {"31", "thirty-first"},
+                {"39", "thirty-ninth"},
+                {"100", "one hundredth"},
+                {"101", "one hundred first"},
+                {"200", "two hundredths"},
+                {"201", "two hundreds first"},
+                {"300", "three hundredths"},
+                {"301", "three hundreds first"},
```

Comment:
Obviously, this looks weird in English, but I presume you've got exactly this pattern in other languages?

Reply:
Yes several languages have this grammar style for large ordinals, and they’re currently unable to support them without huge amounts of copied and tweaked rules. Yes, I agree that this is contrived for English, but it’s readable. It also doesn’t require extensive research on the correct spelling of ordinals in an affected language.

Reply:
Fine with me; just trying to understand what you're doing.